src/librustc_mir/build/matches/mod.rs

```diff
@@ -781,87 +829,172 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
-                debug!(
-                    "match_candidates: add false edges for unreachable {:?} and unmatched {:?}",
-                    unreachable_candidates, unmatched_candidates
-                );
-                for candidate in unreachable_candidates {
-                    let source_info = self.source_info(candidate.span);
-                    let target = self.cfg.start_new_block();
-                    if let Some(otherwise) =
-                        self.bind_and_guard_matched_candidate(target, arm_blocks, candidate)
-                    {
-                        self.cfg
-                            .terminate(otherwise, source_info, TerminatorKind::Unreachable);
-                    }
-                }
+        let first_candidate = &reachable_candidates[0];
 
-                if unmatched_candidates.is_empty() {
-                    return vec![];
+        let candidate_source_info = self.source_info(first_candidate.span);
+
+        self.cfg.terminate(
+            block,
+            candidate_source_info,
+            TerminatorKind::Goto {
+                target: first_candidate.pre_binding_block,
+            },
+        );
+
+        for window in reachable_candidates.windows(2) {
+            if let [first_candidate, second_candidate] = window {
+                let source_info = self.source_info(first_candidate.span);
+                if let Some(otherwise_block) = first_candidate.otherwise_block {
+                    self.cfg.terminate(
+                        otherwise_block,
+                        source_info,
+                        TerminatorKind::FalseEdges {
+                            real_target: second_candidate.pre_binding_block,
+                            imaginary_targets: vec![
+                                first_candidate.next_candidate_pre_binding_block
+                            ],
+                        }
+                    )
                 } else {
-                    let target = self.cfg.start_new_block();
-                    return self.match_candidates(
-                        span,
-                        arm_blocks,
-                        unmatched_candidates,
-                        target,
-                        &mut None,
-                    );
+                    bug!("candidate other than the last has no guard");
                 }
+            } else {
+                bug!("<[_]>::windows returned incorrectly sized window");
             }
         }
 
-        // If there are no candidates that still need testing, we're done.
-        // Since all matches are exhaustive, execution should never reach this point.
-        if unmatched_candidates.is_empty() {
-            return vec![block];
-        }
 
-        // Test candidates where possible.
-        let (otherwise, tested_candidates) =
-            self.test_candidates(span, arm_blocks, &unmatched_candidates, block, fake_borrows);
+        // if None is returned, then
```

Comment:
Was this left over from the old " If there are no candidates that still need testing, we're done." comment? It seems like it does not add anything and in fact I cannot tell which `None` it might be referencing...